### PR TITLE
Add developer environment

### PR DESCRIPTION
**Pull request by vsoch**
hey @SteVwonder! I thought it would be fun to try building this, and possibly seeing if I could continue the work that you started a few years ago. My first step was to add a Developer Environment (included here) and ensure that I could do a basic build. I also saw that your test integration branch was merged into flux-rs, so I updated that, along with one suggestion from cargo.

If you are able, aside from a quick review here, could you give me a high level summary of what you wanted to do here, and other ideas for Flux + rust? I'd like to get better at Rust (and haven't found a good opportunity) and am hoping this might be it. If there is even a small "next step" I could work on somewhere it would be hugely fun and appreciated.

Happy weekend!! :mountain:

### CODE REVIEW

**Comment by SteVwonder** [comment]
Hey @vsoch! I used this repo for similar reasons (an excuse to use rust in my day job). The high level idea was to implement a module for flux that could provide the "OpenMP-style" dependencies defined in this RFC: https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_26.html. The summary motivation for those type of dependencies is they let you specify a DAG of dependencies without (as a user) having to deal with the job ids - so your higher-level workflow system logic is simpler and your dependency specification is identical across workflow runs. And then from a system/Flux-perspective, they are nice because they enable to be more stateless. You don't have to keep a history of all jobs from all time nor due any DB lookups (e.g., a `afterok:<JOBID>` requires looking up the completion status of `<JOBID>` which may have finished days ago).

The idea was that a dependency module would be simple enough to complete in a hackathon (or two) and not too much effort wasted if the rust stuff didn't work.

A lot of the implementation details are fuzzy but I remember fighting a lot with the borrow checker trying to build a dependency graph. Specifically, I wanted to use a graph library and embed the job struct a within the graph. I was struggling (due to my n00b rust skillz) with how to make the borrow checker happy with all the pointers. No amount of box and rc seemed to help me. I think I ended up using multiple data structures to avoid this. One was the graph and it used trivially copy-able types (either ints or strings) as the data and then a map that could be used to lookup the actual job info (the key of the map was the data in the graph IIRC). I think that implementation could definitely be improved. 

As far as other ideas for Flux + rust, the way we were at least talking about it was similar to the kernel: get the infrastructure in place so you can use it idiomatically (in both directions - idiomatic for Rust and idiomatic for Flux) and productively, then start implementing non-critical path plugins/modules in it. During my grad school research, I used and abused pymod quite a bit to get POCs up and running quickly. A nice next step on the road to productizing from Python could be rust. 

Now that I'm writing this out, something we never tried but should be possible is writing some CLI commands in rust. Rather than re-writing flux-tree (for example) in C, maybe rust is a good option. Idk if a "flux-top" TUI was ever made, but maybe another good option. 

I think the key is finding something that can live off in its own repo and not cause a lot of gnarly dependency for the other repos.

CC @trws and @cmoussa1, who were partners in crime with Flux + Rust

File: src/main.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
-use log::{error, info, warn};
+// use log::{error, info, warn};
```
Comment (SteVwonder) [inline src/main.rs:1]: Nit: you can probably just remove this entirely if you want

**Review by SteVwonder — APPROVED** [review]
I didn't try and run locally but this LGTM! Cool to see rust being picked up again for Flux!

**Comment by vsoch** [comment]
> The idea was that a dependency module would be simple enough to complete in a hackathon (or two) and not too much effort wasted if the rust stuff didn't work.

Oh nice! So to put together Flux + the module idea, you would use this via `flux module load <path to binary>` or something else? If I can see how it loads / works that will hugely help with my understanding.

> A lot of the implementation details are fuzzy but I remember fighting a lot with the borrow checker trying to build a dependency graph. Specifically, I wanted to use a graph library and embed the job struct a within the graph. I was struggling (due to my n00b rust skillz) with how to make the borrow checker happy with all the pointers. No amount of box and rc seemed to help me. I think I ended up using multiple data structures to avoid this. One was the graph and it used trivially copy-able types (either ints or strings) as the data and then a map that could be used to lookup the actual job info (the key of the map was the data in the graph IIRC). I think that implementation could definitely be improved.

Noted! I can't promise anything (I am likely much more rust noob) but at a high level I understand the issue you are describing.

> As far as other ideas for Flux + rust, the way we were at least talking about it was similar to the kernel: get the infrastructure in place so you can use it idiomatically (in both directions - idiomatic for Rust and idiomatic for Flux) and productively, then start implementing non-critical path plugins/modules in it. 

I'm assuming these aren't broker modules? From [last week](https://github.com/flux-framework/flux-core/issues/5214#issuecomment-1567166019) I learned they run in the broker address space (and it can get messy to implement them externally and then have conflicts).

> Now that I'm writing this out, something we never tried but should be possible is writing some CLI commands in rust. Rather than re-writing flux-tree (for example) in C, maybe rust is a good option. Idk if a "flux-top" TUI was ever made, but maybe another good option.

This might be a good start for me, actually, since this is fairly conceptually similar (I did submit and keygen examples in Go so I sort of have a sense of how it works).

> I think the key is finding something that can live off in its own repo and not cause a lot of gnarly dependency for the other repos.

I totally agree! :raised_hands: 

> CC @trws and @cmoussa1, who were partners in crime with Flux + Rust

I'll wait to hear from them if we want to merge here. Thanks for the quick feedback and background!

**Comment by trws** [comment]
The goal actually was to write broker modules in it, like flux-depend.  It never got 100% complete, but we had something that produced type-safe flux RPC/message bindings that work with rust's async/.await syntax and managed the message lifetimes and responses mostly automatically.  As in, instead of having to manually track all of that stuff, and know the right dynamic type and cast to it, if you wrote the wrong type in the rust code you would get a compile-time error.  That was part of the value proposition really, making it easier to conform to the stringent requirements of code that has to run in something that really can't be allowed to die.  One issue at the time was the abort on allocation failure issue, but we could actually fix that now with a module flag and some wrappers, so... yeah that could be pretty darned cool. As far as I know the most annoying part remaining is probably dealing with bindgen packaging, which needs an LLVM but doesn't really provide a way to get the right one other than "user should figure that out" which is not very satisfying.

**Comment by trws** [comment]
@SteVwonder btw, flux top was made, and it's amazing. Definitely check it out if you find some time to poke around, major props to the team on that one.

Also, the CLI apps in rust idea is a good one, or anywhere we want/need to migrate python code away from python.  [PyO3](https://github.com/PyO3/pyo3) is amazing, and makes it easier to write a python extension module in rust than in anything other than python IMO, it handles almost all the refcount nastiness by leveraging the borrow checker, and the glue nastiness with compile-time codegen.  In fact an example of this popped up [on hacker news](https://www.jelmer.uk/port-py-to-rust.html) yesterday.

**Comment by vsoch** [comment]
@trws ok with you to merge this environment in?

**Pull request merged**